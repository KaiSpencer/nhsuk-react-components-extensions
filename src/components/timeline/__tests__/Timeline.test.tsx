import React from 'react';
import { shallow } from 'enzyme';
import Timeline from '../Timeline';
import Event, { EventProps } from '../components/Event';
import Tag from '../../tag/Tag';

const actionLinkText = 'Cancel';
const eventInstigator = 'System';

const testEvents: EventProps[] = [
  {
    title: <>Result sent</>,
    instigator: <>{eventInstigator}</>,
    date: '19 Nov 2019, 7:49:10 pm',
    description: [
      <>Test Result: (39S) Low-grade dyskaryosis, HPV positive, Repeat Advised</>,
      <>Test Date: 19-Oct-2020, 9:00:00 am</>,
    ],
    action: (
      <span className="nhsuk-timeline__action">
        -
        <Tag className="nhsuk-timeline__tag" color="yellow">
          Send to printer
        </Tag>
        <a className="nhsuk-timeline__link" href="/placeholder/">
          Cancel
        </a>
      </span>
    ),
  },
  {
    title: <>Patient deferred</>,
    date: '19 Nov 2019, 4:28:57 pm',
    description: [
      <>Defer Reason: Pregnancy</>,
      <>CRM Case Number: CAS-12345-ABCDE</>,
      <>Comments: Requested via the GP form</>,
    ],
  },
  {
    title: <>Screening invitation sent</>,
    instigator: <>System</>,
    date: '18 Nov 2019, 7:49:10 pm',
    action: (
      <span className="nhsuk-timeline__action">
        -
        <Tag className="nhsuk-timeline__tag" color="grey">
          Sent to patient
        </Tag>
        <a className="nhsuk-timeline__link" href="/placeholder2">
          Resend
        </a>
      </span>
    ),
  },
];

describe('Timeline', () => {
  it('matches snapshot', () => {
    const timeline = shallow(<Timeline events={testEvents} />);

    expect(timeline).toMatchSnapshot();

    timeline.unmount();
  });

  it('should have correct classes applied', () => {
    const timeline = shallow(<Timeline events={testEvents} />);

    expect(timeline.hasClass('nhsuk-timeline')).toBeTruthy();

    timeline.unmount();
  });

  describe('Event', () => {
    it('matches snapshot', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event).toMatchSnapshot();

      event.unmount();
    });

    it('should have correct classes applied', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event.hasClass('nhsuk-timeline__event')).toBeTruthy();

      event.unmount();
    });

    it('should have correct title', () => {
      const event = shallow(<Event {...testEvents[1]} />);

      expect(event.find('.nhsuk-timeline__title').text()).toBe('Patient deferred by James Smith');

      event.unmount();
    });

    it('should have correct instigator if instigator data is present', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event.find('.nhsuk-timeline__by').exists()).toBeTruthy();
      expect(event.find('.nhsuk-timeline__by').text()).toBe(` by ${eventInstigator}`);

      event.unmount();
    });

    it('should not have instigator if instigator data is not present', () => {
      const event = shallow(<Event {...testEvents[1]} />);

      expect(event.find('.nhsuk-timeline__by').exists()).toBeFalsy();

      event.unmount();
    });

    it('should have correct date', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event.find('.nhsuk-timeline__date time').text()).toBe(testEvents[0].date);

      event.unmount();
    });

    it('has correct description items if description data is present', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event.find('.nhsuk-timeline__description-item').exists()).toBeTruthy();
    });

    it('has no description items if description data is not present', () => {
      const event = shallow(<Event {...testEvents[2]} />);

      expect(event.find('.nhsuk-timeline__description-item').exists()).toBeFalsy();
    });

    it('has correct link text if action data is present', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event.find('.nhsuk-timeline__link').text()).toBe(actionLinkText);
    });

    it('will show action if action data present', () => {
      const event = shallow(<Event {...testEvents[0]} />);

      expect(event.find('.nhsuk-timeline__action').exists()).toBeTruthy();
    });

    it('will not show action if no action data present', () => {
      const event = shallow(<Event {...testEvents[1]} />);

      expect(event.find('.nhsuk-timeline__action').exists()).toBeFalsy();
    });
  });
});
