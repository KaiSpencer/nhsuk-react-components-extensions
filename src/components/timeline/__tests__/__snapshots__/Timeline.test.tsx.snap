// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Event matches snapshot 1`] = `
<div
  className="nhsuk-timeline__event nhsuk-u-margin-bottom-2"
>
  <h2
    className="nhsuk-timeline__title"
  >
    <span
      className="nhsuk-timeline__status"
    >
      Result sent
    </span>
    <span
      className="nhsuk-timeline__by"
    >
       by 
      System
    </span>
    <span
      className="nhsuk-timeline__action"
    >
      <span
        className="nhsuk-timeline__action"
      >
        -
        <Tag
          className="nhsuk-timeline__tag"
          color="yellow"
        >
          Send to printer
        </Tag>
        <a
          className="nhsuk-timeline__link"
          href="/placeholder/"
        >
          Cancel
        </a>
      </span>
    </span>
  </h2>
  <p
    className="nhsuk-timeline__date"
  >
    <time
      dateTime="19 Nov 2019, 7:49:10 pm"
    >
      19 Nov 2019, 7:49:10 pm
    </time>
  </p>
  <p
    className="nhsuk-timeline__description-item"
    key="0"
  >
    Test Result: (39S) Low-grade dyskaryosis, HPV positive, Repeat Advised
  </p>
  <p
    className="nhsuk-timeline__description-item"
    key="1"
  >
    Test Date: 19-Oct-2020, 9:00:00 am
  </p>
</div>
`;

exports[`Timeline matches snapshot 1`] = `
<div
  className="nhsuk-timeline"
>
  <Event
    action={
      <span
        className="nhsuk-timeline__action"
      >
        -
        <Tag
          className="nhsuk-timeline__tag"
          color="yellow"
        >
          Send to printer
        </Tag>
        <a
          className="nhsuk-timeline__link"
          href="/placeholder/"
        >
          Cancel
        </a>
      </span>
    }
    date="19 Nov 2019, 7:49:10 pm"
    description={
      Array [
        <React.Fragment>
          Test Result: (39S) Low-grade dyskaryosis, HPV positive, Repeat Advised
        </React.Fragment>,
        <React.Fragment>
          Test Date: 19-Oct-2020, 9:00:00 am
        </React.Fragment>,
      ]
    }
    instigator={
      <React.Fragment>
        System
      </React.Fragment>
    }
    key="0"
    title={
      <React.Fragment>
        Result sent
      </React.Fragment>
    }
  />
  <Event
    date="19 Nov 2019, 4:28:57 pm"
    description={
      Array [
        <React.Fragment>
          Defer Reason: Pregnancy
        </React.Fragment>,
        <React.Fragment>
          CRM Case Number: CAS-12345-ABCDE
        </React.Fragment>,
        <React.Fragment>
          Comments: Requested via the GP form
        </React.Fragment>,
      ]
    }
    instigator={
      <React.Fragment>
        James Smith
      </React.Fragment>
    }
    key="1"
    title={
      <React.Fragment>
        Patient deferred
      </React.Fragment>
    }
  />
  <Event
    action={
      <span
        className="nhsuk-timeline__action"
      >
        -
        <Tag
          className="nhsuk-timeline__tag"
          color="grey"
        >
          Sent to patient
        </Tag>
        <a
          className="nhsuk-timeline__link"
          href="/placeholder2"
        >
          Resend
        </a>
      </span>
    }
    date="18 Nov 2019, 7:49:10 pm"
    instigator={
      <React.Fragment>
        System
      </React.Fragment>
    }
    key="2"
    title={
      <React.Fragment>
        Screening invitation sent
      </React.Fragment>
    }
  />
</div>
`;
